private/bufpkg/bufimage/import_tracker.go

```diff
@@ -0,0 +1,226 @@
+	t.findUsedImportsInOptions(file, enum.Options)
+	for _, value := range enum.Value {
+		t.findUsedImportsInOptions(file, value.Options)
+	}
+}
+
+func (t *importTracker) findUsedImportsInField(file *imagev1.ImageFile, field *descriptorpb.FieldDescriptorProto) {
+	t.findUsedImportsInOptions(file, field.Options)
+	switch field.GetType() {
+	case descriptorpb.FieldDescriptorProto_TYPE_MESSAGE,
+		descriptorpb.FieldDescriptorProto_TYPE_GROUP,
+		descriptorpb.FieldDescriptorProto_TYPE_ENUM:
+		t.markUsed(file, field.GetTypeName())
+	}
+	extendee := field.GetExtendee()
+	if extendee != "" {
+		t.markUsed(file, extendee)
+	}
+}
+
+func (t *importTracker) findUsedImportsInOptions(file *imagev1.ImageFile, optionMessage proto.Message) {
+	optionMessage.ProtoReflect().Range(func(field protoreflect.FieldDescriptor, val protoreflect.Value) bool {
+		if field.IsExtension() {
+			t.markUsed(file, string(field.FullName()))
+		}
+		t.findUsedImportsInOptionValue(file, field, val)
+		return true
+	})
+}
+
+func (t *importTracker) findUsedImportsInOptionValue(file *imagev1.ImageFile, optionField protoreflect.FieldDescriptor, val protoreflect.Value) {
+	switch {
+	case optionField.IsMap():
+		if optionField.MapValue().Message() == nil {
+			return // no messages to examine
+		}
+		val.Map().Range(func(k protoreflect.MapKey, v protoreflect.Value) bool {
+			t.findUsedImportsInMessageValue(file, v.Message())
+			return true
+		})
+	case optionField.IsList():
+		if optionField.Message() == nil {
+			return // no messages to examine
+		}
+		list := val.List()
+		for i, l := 0, list.Len(); i < l; i++ {
+			t.findUsedImportsInMessageValue(file, list.Get(i).Message())
+		}
+	case optionField.Message() != nil:
+		t.findUsedImportsInMessageValue(file, val.Message())
+	}
+}
+
+func (t *importTracker) findUsedImportsInMessageValue(file *imagev1.ImageFile, msg protoreflect.Message) {
+	if msg.Descriptor().FullName() == anyMessageName {
+		typeURLField := msg.Descriptor().Fields().ByNumber(1)
+		if typeURLField == nil || typeURLField.Kind() != protoreflect.StringKind || typeURLField.IsList() {
+			// ruh-roh... this should not happen
+			return
+		}
+		valueField := msg.Descriptor().Fields().ByNumber(2)
+		if valueField == nil || valueField.Kind() != protoreflect.BytesKind || valueField.IsList() {
+			// oof, this should not happen
+			return
+		}
+
+		typeURL := msg.Get(typeURLField).String()
+		msgType, err := t.resolver.FindMessageByURL(typeURL)
+		if err != nil {
+			// message is not present in the image
+			return
+		}
+		t.markUsed(file, string(msgType.Descriptor().FullName()))
+		// process Any messages that might be nested inside this one
+		value := msg.Get(valueField).Bytes()
+		nestedMessage := msgType.New()
+		err = proto.UnmarshalOptions{Resolver: t.resolver}.Unmarshal(value, nestedMessage.Interface())
+		if err != nil {
+			// bytes are not valid; skip it
+			return
+		}
+		msg = nestedMessage // fall-through to recurse into this message
```

Comment:
I don't completely follow why we recurse based on the message value here. In what case is this required?

This is probably the analog of `findUsedImportsInOptionValue`, but for Any instead of extensions. However, if you find a concrete value of an any or an extension and you can find the file in which its defined, isn't that enough for marking files/types as imported?

Reply:
Ohh thinking about this more does this capture the case of
```
import "a.proto";
import "b.proto";
import "c.proto";
string x = 1 [(some.ext) = {[type.googleapi.com/some.Foo]: {[some.other.ext]: true}}];
```
where
```
// a.proto
pkg some;
extends google.protobuf.FieldOptions {
    google.protobuf.Any ext = ....;
}
// b.proto
import "d.proto";
pkg some;
message Foo {
   Dmessage x = 1;
   extensions 10 to 100;
}
// c.proto
pkg some.other;
extends some.Foo {
    bool ext = 50;
}
// d.proto
pkg some;
message Dmessage{}
```
so that c is marked as used?

Reply:
Yes, exactly. Also, it's not _just_ extensions that can be referenced this way: an `Any` could be nested inside another `Any`. It's even verified in the test cases:
https://github.com/bufbuild/buf/pull/1835/files#diff-677a988d530972605d439fb3b874951605fee6831d05a32a6e9c456630815a41R19-R29

This is also why I added a TODO to bufimageutil to eventually do the same thing.